react/features/face-centering/middleware.js

```diff
@@ -0,0 +1,135 @@
+// @flow
+
+import {
+    CONFERENCE_JOINED,
+    CONFERENCE_WILL_LEAVE,
+    getCurrentConference
+} from '../base/conference';
+import { JitsiConferenceEvents } from '../base/lib-jitsi-meet';
+import { MiddlewareRegistry } from '../base/redux';
+import { TRACK_UPDATED, TRACK_ADDED, TRACK_REMOVED } from '../base/tracks';
+import { changeTrack, resetTrack } from '../facial-recognition';
+import { shouldDisplayTileView } from '../video-layout';
+import { SET_TILE_VIEW } from '../video-layout/actionTypes';
+import { VIRTUAL_BACKGROUND_TRACK_CHANGED } from '../virtual-background/actionTypes';
+
+import { UPDATE_FACE_COORDINATES } from './actionTypes';
+import {
+    loadWorker,
+    stopFaceRecognition,
+    startFaceRecognition
+} from './actions';
+import { FACE_BOX_EVENT_TYPE } from './constants';
+
+MiddlewareRegistry.register(({ dispatch, getState }) => next => action => {
+    const state = getState();
+    const { faceCoordinatesSharing } = state['features/base/config'];
+    const isInTileView = shouldDisplayTileView(state);
+
+    if (!getCurrentConference(getState())) {
+        return next(action);
+    }
+
+    if (action.type === CONFERENCE_JOINED) {
+        if (faceCoordinatesSharing?.enabled) {
+            dispatch(loadWorker());
+
+            if (isInTileView) {
+                dispatch(startFaceRecognition());
+            }
+        }
+
+        // allow using remote face centering data when local face centering is not enabled
+        action.conference.on(
+            JitsiConferenceEvents.ENDPOINT_MESSAGE_RECEIVED,
+            (...args) => {
+                if (args && args.length >= 2) {
+                    const [ { _id }, eventData ] = args;
+
+                    if (eventData.type === FACE_BOX_EVENT_TYPE) {
+                        dispatch({
+                            type: UPDATE_FACE_COORDINATES,
+                            faceBox: eventData.faceBox,
+                            id: _id
+                        });
+                    }
+                }
+            });
+
+        return next(action);
+    }
+
+    if (!faceCoordinatesSharing?.enabled) {
+        return next(action);
+    }
+
+    // might need to change logic here if object-fit: cover will be used outside of tileview in the future
+    if (action.type === SET_TILE_VIEW) {
+        if (action.enabled) {
+            dispatch(startFaceRecognition());
+        } else {
+            dispatch(stopFaceRecognition());
+        }
+
+        return next(action);
+    }
+
+    if (!isInTileView) {
```

Comment:
It seems to me that we are not going to send face coordinates with others unless we are in tile view. Is this correct?

If this is true, and the user is in stage view but all other participants are in tile view, they won't receive any coordinates. Is that true? Is this desired behaviour?

Shouldn't we always send them when we can and  just use them on the receiver side if the receiver is in tile view?

Reply:
very good point, changed it to always send coordinates if sharing is enabled